ovs/lib/alba.py

```diff
@@ -1681,236 +1682,372 @@ def unlink_alba_backends(target_guid, linked_guid):
+
+                    # Remove all services on ALBA Nodes which have not been checked because requested amount has been reached
+                    for alba_node, alba_backend_info in services_per_node.iteritems():
+                        if alba_backend.name in alba_backend_info:
+                            services_to_remove.append([alba_node, alba_backend_info[alba_backend.name], 'Too many services'])
+                            load_per_node[alba_node] -= 1
+
+            # Always make sure to have at least 1 service for each ALBA Backend, even if it has no OSDs yet
+            if services_for_backend == 0:
+                min_load = min(load_per_node.values()) if load_per_node else 0
+                AlbaController._logger.debug('No services yet for ALBA Backend {0}. Minimum load is {1}'.format(alba_backend.name, min_load))
+                for alba_node, load in load_per_node.iteritems():
+                    if load == min_load:
+                        services_to_add.append([alba_node, 'Minimal load detected'])
+                        load_per_node[alba_node] += 1
+                        break
+
+            if services_to_remove:
+                AlbaController._logger.info('Removing {0} service{1}'.format(len(services_to_remove), '' if len(services_to_remove) == 1 else 's'))
+            for alba_node, service_name, reason in services_to_remove:
                 # noinspection PyTypeChecker
-                if _add_service(node, service_name, alba_backend):
-                    made_changes = True
-                    AlbaController._logger.info('* Added service {0} on {1}: OK'.format(service_name, node.ip))
-                else:
-                    AlbaController._logger.warning('* Added service {0} on {1}: FAIL'.format(service_name, node.ip))
-            if made_changes is True:
-                alba_backend.invalidate_dynamics(['live_status'])
-                alba_backend.backend.invalidate_dynamics(['live_status'])
+                success_remove &= remove_service(_alba_backend=alba_backend, _alba_node=alba_node, _service_name=service_name, _reason=reason)
+
+            if services_to_add:
+                AlbaController._logger.info('Adding {0} service{1}'.format(len(services_to_add), '' if len(services_to_add) == 1 else 's'))
+            for alba_node, reason in services_to_add:
+                # noinspection PyTypeChecker
+                success_add &= add_service(_alba_backend=alba_backend, _alba_node=alba_node, _reason=reason)
+            if len(services_to_remove) > 0 or len(services_to_add) > 0:
+                alba_backend.invalidate_dynamics('live_status')
+                alba_backend.backend.invalidate_dynamics('live_status')
 
-            AlbaController._logger.info('Finished service work log for {0}'.format(name))
+        if success_add is False or success_remove is False:
+            raise Exception('Maintenance agent checkup was not completely successful')
```

Comment:
Raise a custom exception so we could anticipate a failure
Add the raises to the docstring

Reply:
Don't see the added value for a custom exception here
Nobody checks these scheduled tasks at the moment